apollo-router-core/src/cache.rs

```diff
@@ -0,0 +1,176 @@
+use crate::{CacheResolver, CacheResolverError};
+use futures::lock::Mutex;
+use lru::LruCache;
+use std::cmp::Eq;
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::Hash;
+use tokio::sync::broadcast::{self, Sender};
+
+/// A result from a cache get().
+pub type CacheResult<V> = Result<V, CacheResolverError>;
+
+/// A caching map optimised for slow value resolution.
+///
+/// The CachingMap hold values in an LruCache. Values are loaded into the cache on a cache miss and
+/// the cache relies on the resolver to provide values. There is no way to manually remove, update
+/// or otherwise invalidate a cache value at this time. Values will be evicted from the cache once
+/// the cache_limit is reached.
+pub struct CachingMap<K, V> {
+    cached: Mutex<LruCache<K, CacheResult<V>>>,
+    #[allow(clippy::type_complexity)]
+    wait_map: Mutex<HashMap<K, Sender<(K, CacheResult<V>)>>>,
+    cache_limit: usize,
+    resolver: Box<dyn CacheResolver<K, V> + Send + Sync>,
```

Comment:
This dynamic dispatching is unnecessary. We should probably use a generic parameter instead.

Reply:
I put a comment in the PR about how the resolver would be used. It made the code a lot cleaner and I really don't think the performance matters in this use case.

Reply:
Ah sorry I didn't read.

Reply:
> After spending a bit of time exploring different ideas, I've settled
> on the idea proposed by Geoffroy to provide a value resolver for use
> by the cache map.
> 
> This approach is cleanest from the point of view of client interactions,
> although it does require clients to implement the CacheResolver trait.

You don't really explain why this approach was better. In my opinion it is much better if the user just provide a function-like.

> Ideally we could store a callback in the CachingMap struct, but
> lifetimes and async restrictions mean I can't make that work at the
> moment.

Is this comment related?

What about this?

```rust
use std::marker::PhantomData;

pub struct CachingStuff<K, V, F> {
    callback: F,
    phantom_data: PhantomData<(K, V)>,
}

impl<K, V, F> CachingStuff<K, V, F>
where
    F: Fn(K) -> V,
{
    pub fn new(callback: F) -> Self {
        Self {
            callback,
            phantom_data: PhantomData,
        }
    }

    pub fn get(&self, key: K) -> V {
        (self.callback)(key)
    }
}

#[derive(Debug)]
pub struct MyError;

fn main() {
    let cache = CachingStuff::new(|_key: String| Ok::<u32, MyError>(42));
    let value_res: Result<u32, MyError> = cache.get("boo".to_string());
    println!("Value: {:?}", value_res);
}
```

Reply:
Sorry if the detail isn't apparent. The issue was caused by trying to maintain an async function callback. The various things we are calling are async and I wanted to provide an async interface to prevent blocking of threads when retrieving values (I guess I could have used spawn_blocking with a sync interface...).

I couldn't figure out a nice way to store such a "thing" (async fn callback effectively) in a struct.

Reply:
```rust
use std::future::Future;
use std::marker::PhantomData;

pub struct CachingStuff<K, V, F, FUT> {
    callback: F,
    phantom_data: PhantomData<(K, V, FUT)>,
}

impl<K, V, F, FUT> CachingStuff<K, V, F, FUT>
where
    F: Fn(K) -> FUT,
    FUT: Future<Output = V>,
{
    pub fn new(callback: F) -> Self {
        Self {
            callback,
            phantom_data: PhantomData,
        }
    }

    pub async fn get(&self, key: K) -> V {
        (self.callback)(key).await
    }
}

#[derive(Debug)]
pub struct MyError;

#[tokio::main]
async fn main() {
    let cache = CachingStuff::new(|_key: String| async { Ok::<u32, MyError>(42) });
    let value_res: Result<u32, MyError> = cache.get("boo".to_string()).await;
    println!("Value: {:?}", value_res);
    let cache = CachingStuff::new(|_key: String| async {
        tokio::task::spawn_blocking(move || Ok::<u32, MyError>(42))
            .await
            .unwrap()
    });
    let value_res: Result<u32, MyError> = cache.get("boo".to_string()).await;
    println!("Value: {:?}", value_res);
}
```

Reply:
I'd rather we have the trait and a generic member, implement the trait over Fn, and let the calling side decide how they will use it. This will change nothing over the functionality, be more flexible and you'll have the function if you want

Reply:
It's make the code more complicated for no good reason.

This cache can only call one function per caching thing anyway. If we have new use cases and it feels relevant to add a trait, then we can add a trait. Until then please keep the code stupid simple.

Reply:
This discussion has been extracted and deferred into issue: #244